File: cube.js
module.exports = {
  // Mapping of security contexts to data model versions
  // https://cube.dev/docs/reference/configuration/config#contexttoappid
  contextToAppId: ({
    securityContext
  }) => {
    //return securityContext.team;
    return securityContext;
  },
  // Security hook which is run before a query is executed
  // https://cube.dev/docs/reference/configuration/config#queryrewrite
  // queryRewrite: (query, {
  //   securityContext
  // }) => {
  //   if (!securityContext.team) {
  //     securityContext.team = 'public';
  //   }
  //   return query;
  // },
  checkSqlAuth: (query, username) => {
    const securityContext = {
      team: username
    };
    return {
      password: process.env.CUBEJS_SQL_PASSWORD,
      securityContext: securityContext
    };
  },
  semanticLayerSync: () => {
    return [{
      type: "preset",
      name: "828 Sync",
      config: {
        database: "828 Take 2",
        api_token: "448d9518-3d7a-41e3-9bda-e626b1646c32",
        api_secret: "e5e5374a64e12b0f7907e417f0d93ad11e7cbe0046181ba1896671d9949c7120",
        workspace_url: "5276833d.us2a.app.preset.io"
      }
    }, {
      type: "tableau",
      name: "Tableau Sync",
      config: {
        database: "Cube Cloud",
        region: "10ax",
        site: "tonycube",
        personalAccessToken: "cube demo",
        personalAccessTokenSecret: process.env.CUBEJS_TABLEAU_PAT_SECRET
      }
    }];
  }
};

File: cube.py
from cube import config
import os

@config('can_switch_sql_user')
def can_switch_sql_user(current_user: str, new_user: str) -> dict:
  return True

@config('check_sql_auth')
def check_sql_auth(req: dict, user_name: str, password: str) -> dict:
  valid_password = os.getenv('CUBEJS_SQL_PASSWORD')
  if password and password != valid_password:
    raise Exception('Access denied')
  return {
    'password': valid_password,
    'securityContext': {
      'user_name': user_name
    }
  }

@config('query_rewrite')
def query_rewrite(query: dict, ctx: dict) -> dict:
  user_name = ctx.get('securityContext', {}).get('user_name', 'cube')
  if user_name != "cube": 
    query['filters'].append({
      'member': 'revenue_margin.supplier_region',
      'operator': 'equals',
      'values': regions.get(user_name, {}).get('regions', ['AMERICA']),
    })
  return query

regions = {
  'morgan@cube.dev': {'regions': ['EUROPE']},
  'mike@cube.dev': {'regions': ['MIDDLE EAST']}
}
